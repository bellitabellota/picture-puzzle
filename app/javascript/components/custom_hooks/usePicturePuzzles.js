import {useState, useEffect} from "react";

const usePicturePuzzles = () => {
  const [picturePuzzles , setPicturePuzzles ] = useState([]);
  const [error, setError] = useState(null);
 
  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;

    const url = "/api/v1/picture_puzzles"
    
    fetch(url, {signal})
    .then((response) => {
      if(!response.ok) {
        throw new Error(`HTTP Error ${response.status}: ${response.statusText}`);
      }
      return response.json();
    })
    .then((response) => {
      setPicturePuzzles(response);
    }).catch((error) => {
      if (error.name !== "AbortError") {
        setError(error);
      }
    })

    return () => controller.abort(); // the cleanup function was added because the effect runs when the component mounts. So in case the component unmount before the request completes, the request would have been still active but the component would not be there anymore to handle the response.
    
  }, [])

  return { picturePuzzles, error }
}

export default usePicturePuzzles;