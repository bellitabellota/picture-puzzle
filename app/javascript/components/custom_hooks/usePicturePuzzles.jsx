import React, {useState, useEffect} from "react";

const usePicturePuzzles = () => {
  const [picturePuzzles , setPicturePuzzles ] = useState([]);
  const [error, setError] = useState(null);
 
  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;

    const url = "/api/v1/picture_puzzles"
    
    fetch(url, {signal})
    .then((response) => {
      if(!response.ok) {
        throw new Error("Network response was not ok.");
      }
      return response.json();
    })
    .then((response) => {
      setPicturePuzzles(response);
    }).catch(() => {
      setError(error)
    })

    return () => controller.abort();
    
  }, [])

  return { picturePuzzles, error }
}

export default usePicturePuzzles;