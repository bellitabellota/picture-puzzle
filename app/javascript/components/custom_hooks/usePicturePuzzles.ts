import {useState, useEffect} from "react";

import type {PreviewPicturePuzzleType} from "../../types/PreviewPicturePuzzleType"

function isPreviewPicturePuzzle(data: any): data is PreviewPicturePuzzleType {
  return(
    data &&
    typeof data.id === "number" &&
    typeof data.title === "string" &&
    typeof data.imageSrc === "string"
  )
}

function isPreviewPicturePuzzleArray(data: any): data is PreviewPicturePuzzleType[] {
  return(
    Array.isArray(data) && data.every(isPreviewPicturePuzzle)
  )
}

const usePicturePuzzles = () => {
  const [picturePuzzles , setPicturePuzzles ] = useState<PreviewPicturePuzzleType[]>([]);
  const [error, setError] = useState<Error|null>(null);
  const [isLoading, setIsLoading] = useState(true);
 
  useEffect(() => {
    const controller = new AbortController();
    const signal = controller.signal;

    const url = "/api/v1/picture_puzzles"
    
    fetch(url, {signal})
    .then((response) => {
      if(!response.ok) {
        throw new Error(`HTTP Error ${response.status}: ${response.statusText}`);
      }
      return response.json();
    })
    .then((response) => {
      if(!isPreviewPicturePuzzleArray(response)) {
        throw Error("Invalid picture puzzle preview data from API")
      }
      setPicturePuzzles(response);
    }).catch((error) => {
      if (error.name !== "AbortError") {
        setError(error);
      }
    }).finally(() => {
      if (!signal.aborted) {
        setIsLoading(false);
      }
    })

    return () => controller.abort(); // the cleanup function was added because the effect runs when the component mounts. So in case the component unmount before the request completes, the request would have been still active but the component would not be there anymore to handle the response.
    
  }, [])

  return { picturePuzzles, error, isLoading }
}

export default usePicturePuzzles;