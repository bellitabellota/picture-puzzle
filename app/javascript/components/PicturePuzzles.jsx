import React from "react";
import { Link } from "react-router-dom";
import usePicturePuzzles from "./custom_hooks/usePicturePuzzles";

function PicturePuzzles() {
  const {picturePuzzles, error, isLoading} = usePicturePuzzles();

  if(isLoading) return <p>Puzzles are loading ...</p>
  if(error) return <p>{error.message}</p>;

  const puzzles = picturePuzzles.map((picturePuzzle) => <Link to={`/${picturePuzzle.id}`} key={picturePuzzle.id}>{picturePuzzle.title}<br></br></Link>)

  return(
    <main>
      <h1>Choose a Puzzle</h1>
      {puzzles}
    </main>
  )
}

export default PicturePuzzles;