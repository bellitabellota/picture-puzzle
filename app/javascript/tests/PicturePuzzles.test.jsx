import PicturePuzzles from "../components/PicturePuzzles";
import usePicturePuzzles from "../components/custom_hooks/usePicturePuzzles";
import {render, screen } from "@testing-library/react";
import React from "react";

vi.mock("../components/custom_hooks/usePicturePuzzles", () => ({
  default: vi.fn()
}))

describe("PicturePuzzles", ()=> {
  it("renders 'Puzzles are loading ...' when isLoading is true", () => {
    usePicturePuzzles.mockReturnValue({picturePuzzles: [], error: null, isLoading: true})
    const {container} = render(<PicturePuzzles />)

    expect(container).toMatchSnapshot();
  })

  it("renders error message if usePicturePuzzles returned an Error", () => {
    usePicturePuzzles.mockReturnValue({picturePuzzles: [], error: new Error("server error"), isLoading: false})

    render(<PicturePuzzles />)
    expect(screen.getByText("server error")).toBeInTheDocument();
  })
});