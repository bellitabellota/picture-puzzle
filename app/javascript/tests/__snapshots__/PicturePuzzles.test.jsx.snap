// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PicturePuzzles > renders homepage with picturePuzzles if usePicturePuzzles returned picturePuzzles 1`] = `
<div>
  <main
    class="main-picture-puzzles"
  >
    <h1>
      Choose A Puzzle
    </h1>
    <div
      class="grid-container"
    >
      <a
        data-discover="true"
        href="/1"
      >
        <div
          class="picture-puzzle"
        >
          <div
            class="image-container"
          >
            <img
              class="puzzle-image"
              src="/picture-puzzle-images/test-puzzle-1.jpg"
            />
          </div>
          <p
            class="puzzle-title"
          >
            Test Puzzle 1
          </p>
        </div>
      </a>
      <a
        data-discover="true"
        href="/2"
      >
        <div
          class="picture-puzzle"
        >
          <div
            class="image-container"
          >
            <img
              class="puzzle-image"
              src="/picture-puzzle-images/test-puzzle-2.jpg"
            />
          </div>
          <p
            class="puzzle-title"
          >
            Test Puzzle 2
          </p>
        </div>
      </a>
    </div>
  </main>
</div>
`;
